### PR TITLE
feat: exception chaining for notify and sync service

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**

1. Add exception chaining to notification and sync service.
2. Replace null aware operator (!) with ? in notify_request_transformer.

**- How I did it**

1. In Sync service, if the exception caught is AtException, call getTraceMessage method and log the error message.
2. In notification service, pass the AtException object to error callback (This is already in place).

**- Description for the changelog**

Chain exceptions occurring in notification service in sync service.

### CODE REVIEW

**Comment by gkc** [comment]
@kalluriramkumar still marked as draft - is it ready to review?

**Comment by sitaram-kalluri** [comment]
> @kalluriramkumar still marked as draft - is it ready to review?

@gkc : It is ready for review. I marked as draft because, at the moment writing unit tests is not feasible for the exception chaining changes. It requires a lot of changes in various places. May i know if this is fine for now or refactor the code to include the unit tests.

**Comment by gkc** [comment]
> @gkc : It is ready for review. I marked as draft because, at the moment writing unit tests is not feasible for the exception chaining changes. It requires a lot of changes in various places. May i know if this is fine for now or refactor the code to include the unit tests.

What tests would you like to write? Let's figure out if we can get them written without extensive changes.

**Review by gkc — CHANGES_REQUESTED** [review]
> @gkc : It is ready for review. I marked as draft because, at the moment writing unit tests is not feasible for the exception chaining changes. It requires a lot of changes in various places. May i know if this is fine for now or refactor the code to include the unit tests.

What tests would you like to write? Let's figure out if we can get them written without extensive changes.

**Comment by sitaram-kalluri** [comment]
> > @gkc : It is ready for review. I marked as draft because, at the moment writing unit tests is not feasible for the exception chaining changes. It requires a lot of changes in various places. May i know if this is fine for now or refactor the code to include the unit tests.
> 
> What tests would you like to write? Let's figure out if we can get them written without extensive changes.

Tests that cover the following scenarios:
1. Encryption issues that trigger the exception are chained and added to notify error back.
2. Server exception are captured in notify error callback.
3. Exception of sync to local and sync to remote are added to sync error call back

**Comment by gkc** [comment]
> > > @gkc : It is ready for review. I marked as draft because, at the moment writing unit tests is not feasible for the exception chaining changes. It requires a lot of changes in various places. May i know if this is fine for now or refactor the code to include the unit tests.
> > 
> > 
> > What tests would you like to write? Let's figure out if we can get them written without extensive changes.
> 
> Tests that cover the following scenarios:
> 
> 1. Encryption issues that trigger the exception are chained and added to notify error back.
> 2. Server exception are captured in notify error callback.
> 3. Exception of sync to local and sync to remote are added to sync error call back

OK please elaborate on each of the test scenarios and describe the things that make writing the tests unfeasible

**Comment by sitaram-kalluri** [comment]
> > > > @gkc : It is ready for review. I marked as draft because, at the moment writing unit tests is not feasible for the exception chaining changes. It requires a lot of changes in various places. May i know if this is fine for now or refactor the code to include the unit tests.
> > > 
> > > 
> > > What tests would you like to write? Let's figure out if we can get them written without extensive changes.
> > 
> > 
> > Tests that cover the following scenarios:
> > 
> > 1. Encryption issues that trigger the exception are chained and added to notify error back.
> > 2. Server exception are captured in notify error callback.
> > 3. Exception of sync to local and sync to remote are added to sync error call back
> 
> OK please elaborate on each of the test scenarios and describe the things that make writing the tests unfeasible

Common changes:
a. Remove the static nature of `isAtSignExists` method `at_client_validation.dart` to mock the look-up of atsign in the root-server. This method is called in two other methods which has to be changed.

1. Encryption issue: 
       a. Return a null value when looked for encryption public key of a sharedWith atsign which results in the encryption failure 
           when notifying with value.
        b. Return a null value when looked for current atsign private key to decrypt the existing encrypted_shared_key of an 
            atsign.
    Changes required:
         a. In notification_server_impl.dart, in notify method (line no. 211) Create a class level field for `AtKeyEncryptionManager` 
             class to mock 'SharedKeyEncryption.class'
3. Server returning the error changes : Common changes would suffice the for this test.
4. Sync service
     Changes required:
        a. Refactor sync_service_impl.dart to pass the mock remote secondary instance.
        b. Refactor the sync_util.dart to pass the mock commit log instance and/or remove the static nature of method to mock the 
           method implementation.

File: at_client/lib/src/encryption_service/abstract_atkey_encryption.dart [hunk 1]
```diff
@@ -44,13 +44,22 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
       // to cache the encryptedSharedKey.
       await _notifyEncryptedSharedKey(atKey, encryptedSharedKey);
       // Store the sharedKey for future retrieval.
-      // Encrypt the sharedKey with currentAtSignPrivateKey and store it for future use.
+      // Encrypt the sharedKey with currentAtSignPublicKey and store it for future use.
+      String? currentAtSignEncryptionPublicKey;
+      try {
+        currentAtSignEncryptionPublicKey = await AtClientManager.getInstance()
```
Comment (murali-shris) [inline at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:50]: if publickey is not found in local.should we lookup remote?

File: at_client/lib/src/service/sync_service_impl.dart [hunk 2]
```diff
@@ -337,13 +339,17 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+                'exception while updating commit entry for entry:$entry Reason: $cause');
+            return keyInfoList;
           }
         }
       } on Exception catch (e) {
+        var cause = (e is AtException) ? e.getTraceMessage() : e.toString();
         _logger.severe(
-            'exception while syncing batch: ${e.toString()} batch commit entries: $unCommittedEntryList');
+            'exception occurred while syncing batch commit entries: $unCommittedEntryList  Reason: $cause');
+        return keyInfoList;
```
Comment (murali-shris) [inline at_client/lib/src/service/sync_service_impl.dart]: check my latest commit to trunk..we should not return keyInfoList on exception

File: at_client/lib/src/service/sync_service_impl.dart [hunk 3]
```diff
@@ -382,8 +388,10 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
           await _syncLocal(serverCommitEntry);
           keyInfoList.add(keyInfo);
         } on Exception catch (e) {
+          var cause = (e is AtException) ? e.getTraceMessage() : e.toString();
           _logger.severe(
-              'exception syncing entry to local $serverCommitEntry - ${e.toString()}');
+              'exception syncing entry to local $serverCommitEntry - $cause');
+          return keyInfoList;
```
Comment (murali-shris) [inline at_client/lib/src/service/sync_service_impl.dart]: check my latest commit to trunk..we should not return keyInfoList on exception

File: at_client/lib/src/encryption_service/abstract_atkey_encryption.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:50]: The encryption public key of current atsign gets from the .atKeys files and then gets synced to remote secondary. So in ideal scenario, the key is not available in local is highly unlikely.

File: at_client/lib/src/service/sync_service_impl.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/service/sync_service_impl.dart]: Removed the return statement

File: at_client/lib/src/service/sync_service_impl.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/service/sync_service_impl.dart]: Removed the return statement

**Review by murali-shris — APPROVED** [review]

**Comment by gkc** [comment]
> > > > > @gkc : It is ready for review. I marked as draft because, at the moment writing unit tests is not feasible for the exception chaining changes. It requires a lot of changes in various places. May i know if this is fine for now or refactor the code to include the unit tests.
> > > > 
> > > > 
> > > > What tests would you like to write? Let's figure out if we can get them written without extensive changes.
> > > 
> > > 
> > > Tests that cover the following scenarios:
> > > 
> > > 1. Encryption issues that trigger the exception are chained and added to notify error back.
> > > 2. Server exception are captured in notify error callback.
> > > 3. Exception of sync to local and sync to remote are added to sync error call back
> > 
> > 
> > OK please elaborate on each of the test scenarios and describe the things that make writing the tests unfeasible
> 
> Common changes: a. Remove the static nature of `isAtSignExists` method `at_client_validation.dart` to mock the look-up of atsign in the root-server. This method is called in two other methods which has to be changed.
> 
> 1. Encryption issue:
>    a. Return a null value when looked for encryption public key of a sharedWith atsign which results in the encryption failure
>    when notifying with value.
>    b. Return a null value when looked for current atsign private key to decrypt the existing encrypted_shared_key of an
>    atsign.
>    Changes required:
>    a. In notification_server_impl.dart, in notify method (line no. 211) Create a class level field for `AtKeyEncryptionManager`
>    class to mock 'SharedKeyEncryption.class'
> 2. Server returning the error changes : Common changes would suffice the for this test.
> 3. Sync service
>    Changes required:
>    a. Refactor sync_service_impl.dart to pass the mock remote secondary instance.
>    b. Refactor the sync_util.dart to pass the mock commit log instance and/or remove the static nature of method to mock the
>    method implementation.

That sounds like a good plan. Please go ahead and implement!

**Review by gkc — APPROVED** [review]

**Pull request merged**